sysfs/class_infiniband.go

```diff
@@ -0,0 +1,272 @@
+	Counters    InfiniBandCounters
+}
+
+type InfiniBandClass map[string]map[uint]InfiniBandPort
+
+// InfiniBandClassDevices scans /sys/class/infiniband for devices and returns them as a list of names.
+func (fs FS) InfiniBandClassDevices() ([]string, error) {
+	var res []string
+	path := fs.sys.Path(infinibandclassPath)
+
+	devices, err := ioutil.ReadDir(path)
+	if err != nil {
+		return res, fmt.Errorf("cannot access %s dir %s", path, err)
+	}
+
+	for _, deviceDir := range devices {
+		if deviceDir.Mode().IsRegular() {
+			continue
+		}
+		res = append(res, deviceDir.Name())
+	}
+
+	return res, nil
+}
+
+// InfiniBandDevicePorts scans /sys/class/infiniband/<device> for ports and returns them as a list of integers.
+func (fs FS) InfiniBandDevicePorts(device string) ([]uint, error) {
+	var res []uint
+	path := fs.sys.Path(infinibandclassPath, device, "ports")
+
+	ports, err := ioutil.ReadDir(path)
+	if err != nil {
+		return res, fmt.Errorf("cannot access %s dir %s", path, err)
+	}
+
+	for _, portDir := range ports {
+		if portDir.Mode().IsRegular() {
+			continue
+		}
+		port, err := strconv.ParseUint(portDir.Name(), 10, 32)
+		if err != nil {
+			return res, fmt.Errorf("failed to convert %s into uint", portDir.Name())
+		}
+		res = append(res, uint(port))
+	}
+
+	return res, nil
+}
+
+// InfiniBandClass returns info for all ports (port) of all InfiniBand devices (device) read from /sys/class/infiniband/<device>/ports/<port>.
+func (fs FS) InfiniBandClass() (InfiniBandClass, error) {
+	devices, err := fs.InfiniBandClassDevices()
+	if err != nil {
+		return nil, err
+	}
+
+	infiniBandClass := InfiniBandClass{}
+	for _, device := range devices {
+		ports, err := fs.InfiniBandDevicePorts(device)
+		if err != nil {
+			return nil, err
+		}
+
+		infiniBandClass[device] = map[uint]InfiniBandPort{}
+		for _, port := range ports {
+			path := fs.sys.Path(infinibandclassPath, device, "ports", fmt.Sprint(port))
+			portClass, err := infiniBandClass.parseInfiniBandPort(path)
+			if err != nil {
+				return nil, err
+			}
+			portClass.Name = device
+			portClass.Port = port
+			infiniBandClass[device][port] = *portClass
+		}
+	}
+	return infiniBandClass, nil
+}
+
+// Parse InfiniBand state. Expected format: "<id>: <string-representation>"
+func parseState(s string) (uint, string, error) {
+	var id uint
+	var name string
+	parts := strings.Split(s, ":")
+	if len(parts) != 2 {
+		return id, name, fmt.Errorf("failed to split %s into 'ID: NAME'", s)
+	}
+	name = strings.TrimSpace(parts[1])
+	value, err := strconv.ParseUint(strings.TrimSpace(parts[0]), 10, 32)
+	if err != nil {
+		return id, name, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	id = uint(value)
+	return id, name, nil
+}
+
+// Parse rate (example: "100 Gb/sec (4X EDR)") and return it as bytes/second
+func parseRate(s string) (uint64, error) {
+	var rate uint64
+	parts := strings.Split(s, "Gb/sec")
```

Comment:
Is this the only unit possible?

Reply:
Yes - see `rate_show` function in https://elixir.bootlin.com/linux/v5.1.7/source/drivers/infiniband/core/sysfs.c:
```c
	return sprintf(buf, "%d%s Gb/sec (%dX%s)\n",
		       rate / 10, rate % 10 ? ".5" : "",
		       ib_width_enum_to_int(attr.active_width), speed);
```

Reply:
I changed to code to also correctly parse `2.5 Gb/sec` and added a test case for it.